### PR TITLE
Afl++ support, IJON support, Reinstated Dockerfile

### PR DESCRIPTION
**Pull request by mahaloz**
* added support for AFL++
* added support for IJON 
* added support Dockerfile
* new check for empty AFL binary

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+RUN apt-add-repository -y universe
+RUN apt-get install -y \
+	build-essential \
+	gcc-multilib \
+	libtool \
+	automake \
+	autoconf \
+	bison \
+	git \
+    gcc \
```
Comment (twizmwazin) [inline Dockerfile:14]: indenting got messed up a bit here

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ 	debian-archive-keyring \
+        libtool-bin
+
+# for QEMU compiling
+RUN cp /etc/apt/sources.list /etc/apt/sources.list~
+RUN sed -Ei 's/^# deb-src /deb-src /' /etc/apt/sources.list
+RUN apt-get update 
+RUN DEBIAN_FRONTEND=noninteractive apt-get build-dep -y \
+	qemu
+RUN apt-get install -y \
```
Comment (twizmwazin) [inline Dockerfile:25]: Can we try and merge apt installations to the first couple layers?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	python3 \
+	python3-pip 
+
+# PHUZZER & DEPS
+RUN pip3 install git+https://github.com/shellphish/shellphish-afl
```
Comment (twizmwazin) [inline Dockerfile:30]: We may want to consider installing this as a wheel. An old version is on pypi, but we also have a wheel in the wheels repo.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+RUN pip3 install git+https://github.com/shellphish/driller
+RUN pip3 install git+https://github.com/angr/tracer
+# Shellphish-AFL Symlinks
+RUN ln -s /usr/local/bin/afl-cgc /usr/bin/afl-cgc
```
Comment (twizmwazin) [inline Dockerfile:34]: Maybe it is possible to remove these if shellphish-afl can be adjusted to check paths?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+# Install IJON Fuzzer
+RUN mkdir /phuzzers/ && cd /phuzzers && \
+        git clone https://github.com/RUB-SysSec/ijon && \
+        apt-get install clang -y && \
+        cd ijon && make && cd llvm_mode && LLVM_CONFIG=llvm-config-6.0 CC=clang-6.0 make
+
+# Install AFL++
+RUN cd /phuzzers/ && \
+	bash -c "$(wget -O - https://apt.llvm.org/llvm.sh)" && \
```
Comment (twizmwazin) [inline Dockerfile:48]: I think the indentation here is a little off too

File: Dockerfile (hunk 1 above)
Comment (mahaloz) [reply Dockerfile:15]: Fixed

File: Dockerfile (hunk 3 above)
Comment (mahaloz) [reply Dockerfile]: Changed to a wheel.

File: Dockerfile (hunk 4 above)
Comment (mahaloz) [reply Dockerfile]: This I can't remove for now.

File: Dockerfile (hunk 5 above)
Comment (mahaloz) [reply Dockerfile:53]: Fixed

File: bin/fuzzify.sh [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  then
+    echo "No Binary name or path provided."
+    echo "To use this, simply give the program the name of the binary"
+    echo "like so: ./fuzzify ls"
+    echo "It will try to reolve it by looking in your path."
+    exit 1
+fi
+
+# get all the linked libs
+libs=$(ldd $(which $1) | grep so | sed -e '/^[^\t]/ d' | sed -e 's/\t//' | sed -e 's/.*=..//' | sed -e 's/ (0.*)//' | grep -o '^\/.*')
```
Comment (twizmwazin) [inline bin/fuzzify.sh:13]: I think you should be able to combine those sed expressions into one sed invocation, and then just separate the statements with semicolons.

File: phuzzer/phuzzers/afl_ijon.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+from .afl import AFL
+import logging
+import os
+
+l = logging.getLogger("phuzzer.phuzzers.afl")
```
Comment (twizmwazin) [inline phuzzer/phuzzers/afl_ijon.py:5]: I think these days we tend to prefer using `__name__` for logger names.

File: phuzzer/phuzzers/afl_plusplus.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
+from .afl import AFL
+import logging
+import os
+import subprocess
+
+l = logging.getLogger("phuzzer.phuzzers.afl")
```
Comment (twizmwazin) [inline phuzzer/phuzzers/afl_plusplus.py:6]: Same as above

File: phuzzer/phuzzers/afl_ijon.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
+from .afl import AFL
```
Comment (twizmwazin) [inline phuzzer/phuzzers/afl_ijon.py:1]: Ideally for new files and when updating imports, we should try to sort them into three sections: standard lib, third-party, and local, alphabetically sorted within each section.

File: phuzzer/phuzzers/afl_plusplus.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
+from .afl import AFL
```
Comment (twizmwazin) [inline phuzzer/phuzzers/afl_plusplus.py:1]: Same as above.

**Review by twizmwazin — COMMENTED** [review]
Just a bunch of pedantry, otherwise LGTM.

File: bin/fuzzify.sh (hunk 6 above)
Comment (mahaloz) [reply bin/fuzzify.sh:13]: this file actually does not need to be here

Comment (mahaloz) [reply bin/fuzzify.sh:13]: I'll remove it

File: phuzzer/phuzzers/afl_ijon.py (hunk 9 above)
Comment (mahaloz) [reply phuzzer/phuzzers/afl_ijon.py]: fixed

File: phuzzer/phuzzers/afl_plusplus.py (hunk 10 above)
Comment (mahaloz) [reply phuzzer/phuzzers/afl_plusplus.py]: fixed

File: phuzzer/phuzzers/afl_ijon.py (hunk 7 above)
Comment (mahaloz) [reply phuzzer/phuzzers/afl_ijon.py]: fixed, I think.

File: phuzzer/phuzzers/afl_plusplus.py (hunk 8 above)
Comment (mahaloz) [reply phuzzer/phuzzers/afl_plusplus.py]: fixed, I think.

**Pull request merged**